superset/assets/javascripts/explorev2/components/ControlPanelsContainer.jsx

```diff
@@ -17,48 +18,65 @@ const propTypes = {
   form_data: PropTypes.object.isRequired,
   y_axis_zero: PropTypes.any,
   alert: PropTypes.string,
+  exploreState: PropTypes.object.isRequired,
 };
 
 class ControlPanelsContainer extends React.Component {
+  constructor() {
+    super();
+    this.fieldOverrides = this.fieldOverrides.bind(this);
+    this.getFieldData = this.getFieldData.bind(this);
+    this.removeAlert = this.removeAlert.bind(this);
+    this.onChange = this.onChange.bind(this);
+  }
   componentWillMount() {
     const datasource_id = this.props.form_data.datasource;
     const datasource_type = this.props.datasource_type;
     if (datasource_id) {
-      this.props.actions.fetchFieldOptions(datasource_id, datasource_type);
+      this.props.actions.fetchDatasourceMetadata(datasource_id, datasource_type);
     }
   }
-
   componentWillReceiveProps(nextProps) {
     if (nextProps.form_data.datasource !== this.props.form_data.datasource) {
       if (nextProps.form_data.datasource) {
-        this.props.actions.fetchFieldOptions(
+        this.props.actions.fetchDatasourceMetadata(
           nextProps.form_data.datasource, nextProps.datasource_type);
       }
     }
   }
-
-  onChange(name, value, label) {
-    this.props.actions.setFieldValue(this.props.datasource_type, name, value, label);
+  onChange(name, value) {
+    this.props.actions.setFieldValue(this.props.datasource_type, name, value);
+  }
+  getFieldData(fs) {
+    const fieldOverrides = this.fieldOverrides();
+    if (!this.props.fields) {
+      return null;
+    }
+    let fieldData = this.props.fields[fs] || {};
+    if (fieldOverrides.hasOwnProperty(fs)) {
+      const overrideData = fieldOverrides[fs];
+      fieldData = Object.assign({}, fieldData, overrideData);
+    }
+    if (fieldData.mapStateToProps) {
```

Comment:
i think this could be more clear as the following:
- renaming `mapStateToProps` to `getChoices` and only passing in the datasource, rather than the entire state.

```js
if (fieldData.getChoices) {
  Object.assign(fieldData, fieldData.getChoices(this.props.datasource));
}
```